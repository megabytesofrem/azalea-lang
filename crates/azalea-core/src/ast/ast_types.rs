use std::hash::Hash;

use crate::parse::span::Span;

use super::{Expr, Stmt};

#[derive(Debug, Clone)]
pub enum Ty {
    Int,
    Float,
    String,
    Bool,
    Unit,

    /// A hole in the type system
    Any,

    /// Represents a type that we don't know yet, will be resolved later on
    Unresolved,

    /// A type variable is a placeholder for a type that can be replaced with any type.
    /// For example, in the type `List[A]`, `A` is a type variable.
    Var(String),

    /// A type constructor is a type that is _parameterized_ by other types.
    /// The following are valid type constructors:
    /// - `List[A]`
    /// - `List[Int]`
    ///
    /// This is not a type constructor:
    /// - `String` ‚ùå
    TypeCons(String, Vec<Ty>),

    /// A type constrained by a type class.
    /// An example of a constraint is `Num[A]`, which means that `A` must be a numeric type.
    /// This is used to represent type classes in the type system.
    Constraint(String, Vec<Ty>),

    /// Maps directly to JavaScript arrays
    Array(Box<Ty>),

    /// Allows us to pass functions around as values
    Fn(Box<Function>),

    /// Record and enum types
    Record(Box<Record>),
    Enum(Box<Enum>),

    /// A universally quantified polymorphic type, e.g. `forall a. a -> a`
    ForAll(Vec<String>, Box<Ty>),
}

// Implement PartialEq, Eq and Hash for `Ty` since it's needed in
// typecheck.

impl PartialEq for Ty {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Ty::Var(a), Ty::Var(b)) => a == b,
            (Ty::TypeCons(a, _), Ty::TypeCons(b, _)) => a == b,
            (Ty::Array(a), Ty::Array(b)) => a == b,
            (Ty::Fn(a), Ty::Fn(b)) => a == b,
            (Ty::Record(a), Ty::Record(b)) => a == b,
            (Ty::Enum(a), Ty::Enum(b)) => a == b,

            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
        }
    }
}

impl Eq for Ty {}

impl Hash for Ty {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        core::mem::discriminant(self).hash(state);
    }
}

/// A pair of a name and a type
pub type TypedPair = (String, Ty);

/// Records are semantically equivalent to JavaScript objects
#[derive(Debug, Clone, PartialEq)]
pub struct Record {
    pub name: String,
    pub fields: Vec<(String, Ty)>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Enum {
    pub name: String,
    pub variants: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Function {
    pub name: String,
    pub args: Vec<(String, Ty)>,

    // Generic type parameters, e.g. `fn foo[T](x: T): T`
    pub type_params: Vec<String>,

    pub return_ty: Ty,

    // Optional body, if this is used in a function declaration
    pub body: Option<Vec<Span<Stmt>>>,
    pub body_expr: Option<Box<Span<Expr>>>, // If this is a lambda expression
}

#[derive(Debug, Clone, PartialEq)]
pub struct RecordExpr {
    pub name: String,
    pub fields: Vec<(String, Expr)>,
}

impl Record {
    pub fn to_type(&self) -> Ty {
        Ty::Record(Box::new(self.clone()))
    }
}

impl Enum {
    pub fn to_type(&self) -> Ty {
        Ty::Enum(Box::new(self.clone()))
    }
}

impl Function {
    pub fn to_type(&self) -> Ty {
        Ty::Fn(Box::new(self.clone()))
    }

    pub fn new_with_empty(name: String, args: Vec<(String, Ty)>, return_ty: Ty) -> Self {
        Self {
            name,
            args,
            type_params: vec![],
            return_ty,
            body: None,
            body_expr: None,
        }
    }

    pub fn new_with_stmts(
        name: String,
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body: Vec<Span<Stmt>>,
    ) -> Self {
        Self {
            name,
            args,
            type_params: vec![],
            return_ty,
            body: Some(body),
            body_expr: None,
        }
    }

    pub fn new_with_expr(
        name: String,
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body_expr: Box<Span<Expr>>,
    ) -> Self {
        Self {
            name,
            args,
            type_params: vec![],
            return_ty,
            body: None,
            body_expr: Some(body_expr),
        }
    }
}
