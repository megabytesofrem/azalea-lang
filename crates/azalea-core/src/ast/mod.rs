pub mod ast;
pub mod ast_types;
pub mod pretty;

// Re-export *everything*
pub use ast::*;
