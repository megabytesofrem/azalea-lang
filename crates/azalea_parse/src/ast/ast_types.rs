use std::hash::Hash;

use crate::span::Span;

use super::{Block, Stmt, ast::Expr};

#[derive(Debug, Clone)]
pub enum Ty {
    Int,
    Float,
    String,
    Bool,
    Unit,

    /// Represents a type that we don't know yet, will be resolved later on
    UnknownForNow,

    Var(String),

    /// An instantiated type. Instantiated types *may* or *may not* be generic.
    /// The following are all instantiated types:
    /// - `string`
    /// - `List[string]`
    /// - `List[A]`: Do note that `A` is a type variable, but for the purposes of simplicity in
    /// our type system we treat it as an Instantiated type.
    Instantiated(String, Vec<Ty>),

    /// Maps directly to JavaScript arrays
    Array(Box<Ty>),

    /// Allows us to pass functions around as values
    Fn(Box<Function>),

    /// Record and enum types
    Record(Box<Record>),
    Enum(Box<Enum>),
}

// Implement PartialEq, Eq and Hash for `Ty` since it's needed in
// typecheck.

impl PartialEq for Ty {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Ty::Var(a), Ty::Var(b)) => a == b,
            (Ty::Instantiated(a, _), Ty::Instantiated(b, _)) => a == b,
            (Ty::Array(a), Ty::Array(b)) => a == b,
            (Ty::Fn(a), Ty::Fn(b)) => a == b,
            (Ty::Record(a), Ty::Record(b)) => a == b,
            (Ty::Enum(a), Ty::Enum(b)) => a == b,

            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
        }
    }
}

impl Eq for Ty {}

impl Hash for Ty {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        core::mem::discriminant(self).hash(state);
    }
}

/// A pair of a name and a type
pub type TypedPair = (String, Ty);

/// Records are semantically equivalent to JavaScript objects
#[derive(Debug, Clone, PartialEq)]
pub struct Record {
    pub name: String,
    pub fields: Vec<(String, Ty)>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Enum {
    pub name: String,
    pub variants: Vec<String>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Function {
    pub name: String,
    pub args: Vec<(String, Ty)>,
    pub return_ty: Ty,

    // Optional body, if this is used in a function declaration
    pub body: Option<Vec<Span<Stmt>>>,
    pub body_expr: Option<Box<Span<Expr>>>, // If this is a lambda expression
}

impl Record {
    pub fn to_type(&self) -> Ty {
        Ty::Record(Box::new(self.clone()))
    }
}

impl Enum {
    pub fn to_type(&self) -> Ty {
        Ty::Enum(Box::new(self.clone()))
    }
}

impl Function {
    pub fn to_type(&self) -> Ty {
        Ty::Fn(Box::new(self.clone()))
    }

    pub fn new_with_empty(name: String, args: Vec<(String, Ty)>, return_ty: Ty) -> Self {
        Self {
            name,
            args,
            return_ty,
            body: None,
            body_expr: None,
        }
    }

    pub fn new_with_stmts(
        name: String,
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body: Vec<Span<Stmt>>,
    ) -> Self {
        Self {
            name,
            args,
            return_ty,
            body: Some(body),
            body_expr: None,
        }
    }

    pub fn new_with_expr(
        name: String,
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body_expr: Box<Span<Expr>>,
    ) -> Self {
        Self {
            name,
            args,
            return_ty,
            body: None,
            body_expr: Some(body_expr),
        }
    }
}
