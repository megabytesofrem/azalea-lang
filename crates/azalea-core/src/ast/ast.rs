use std::fmt;

use super::ast_types::{Enum, Function, Record, RecordExpr, Ty};
use crate::{lexer::Op, parse::span::Span};

/// A block of statements delimited by `do` and `end`
pub type Block = Vec<Span<Stmt>>;

#[derive(Debug, Clone, PartialEq)]
pub enum Literal {
    Int(i32),
    Float(f32),
    String(String),
    Bool(bool),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Member {
    pub target: Box<Span<Expr>>,
    pub name: String,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Stmt {
    /// An expression, used in place of a statement
    Expr(Span<Expr>),

    /// A `let` statement, *not* a binding
    Let {
        name: String,
        ty: Ty,
        value: Box<Span<Expr>>,
    },

    For {
        name: String,
        iterable: Box<Span<Expr>>,
        body: Vec<Span<Stmt>>,
    },

    While {
        cond: Box<Span<Expr>>,
        body: Vec<Span<Stmt>>,
    },

    /// A record declaration
    RecordDecl(Record),

    EnumDecl(Enum),

    /// A function declaration
    FnDecl(Function),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Literal(Literal),
    Ident(String),
    MemberAccess(Member),

    BinOp(Box<Span<Expr>>, Op, Box<Span<Expr>>),
    UnOp(Op, Box<Span<Expr>>),

    Record(RecordExpr),

    Array {
        elements: Vec<Span<Expr>>,
    },

    ArrayIndex {
        target: Box<Span<Expr>>,
        index: Box<Span<Expr>>,
    },

    /// A function call: f(x, y)
    FnCall {
        target: Box<Span<Expr>>,
        args: Vec<Span<Expr>>,
    },

    /// A lambda expression: \(x) -> x + 1
    Lam {
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body: Box<Span<Expr>>,
    },

    If {
        cond: Box<Span<Expr>>,
        then: Vec<Span<Stmt>>,
        else_: Option<Vec<Span<Stmt>>>,
    },
}

impl fmt::Display for Op {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Op::Neg => write!(f, "-"),
            Op::Add => write!(f, "+"),
            Op::Sub => write!(f, "-"),
            Op::Mul => write!(f, "*"),
            Op::Div => write!(f, "/"),
            Op::DoubleEq => write!(f, "=="),
            Op::Not => write!(f, "!"),
            Op::NotEq => write!(f, "!="),
            Op::Less => write!(f, "<"),
            Op::Greater => write!(f, ">"),
            Op::LessEq => write!(f, "<="),
            Op::GreaterEq => write!(f, ">="),
        }
    }
}

impl fmt::Display for Literal {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Literal::Int(i) => write!(f, "{}", i),
            Literal::Float(fl) => write!(f, "{}", fl),
            Literal::String(s) => write!(f, "{}", s),
            Literal::Bool(b) => write!(f, "{}", b),
        }
    }
}

impl fmt::Display for RecordExpr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let fields: Vec<String> = self
            .fields
            .iter()
            .map(|(name, expr)| format!("{}: {}", name, expr))
            .collect();
        write!(f, "{} {{ {} }}", self.name, fields.join(", "))
    }
}

impl fmt::Display for Member {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}.{}", self.target.target, self.name)
    }
}

impl fmt::Display for Expr {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Expr::Literal(lit) => write!(f, "{}", lit),
            Expr::Ident(ident) => write!(f, "{}", ident),
            Expr::MemberAccess(member) => write!(f, "{}", member),
            Expr::BinOp(left, op, right) => {
                write!(f, "({} {} {})", left.target, op, right.target)
            }
            Expr::UnOp(op, expr) => write!(f, "{}({})", op, expr.target),
            Expr::Record(record) => write!(f, "{}", record),
            Expr::Array { elements } => {
                let elems: Vec<String> = elements.iter().map(|e| e.target.to_string()).collect();
                write!(f, "[{}]", elems.join(", "))
            }
            Expr::ArrayIndex { target, index } => write!(f, "{}[{}]", target.target, index.target),
            Expr::FnCall { target, args } => {
                let args_str: Vec<String> = args.iter().map(|a| a.target.to_string()).collect();
                write!(f, "{}({})", target.target, args_str.join(", "))
            }

            Expr::Lam {
                args,
                return_ty,
                body,
            } => {
                let args_str: Vec<String> = args.iter().map(|(n, _)| n.clone()).collect();
                write!(
                    f,
                    "\\({}) -> {} : {}",
                    args_str.join(", "),
                    body.target,
                    return_ty
                )
            }

            Expr::If { cond, then, else_ } => {
                let else_str = if let Some(else_body) = else_ {
                    format!(
                        " else {}",
                        else_body
                            .iter()
                            .map(|s| "<stmt>".to_string()) // Placeholder for actual statement formatting
                            .collect::<Vec<_>>()
                            .join("; ")
                    )
                } else {
                    String::new()
                };
                write!(
                    f,
                    "if {} then do\n{}\nend {}",
                    cond.target,
                    then.iter()
                        .map(|s| "<stmt>".to_string()) // Placeholder for actual statement formatting
                        .collect::<Vec<_>>()
                        .join("; "),
                    else_str
                )
            }
        }
    }
}
