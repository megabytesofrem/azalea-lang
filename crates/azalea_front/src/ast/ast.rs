use super::ast_types::Ty;
use crate::{lexer::Op, span::Span};

pub type Block = Vec<Span<Stmt>>;

#[derive(Debug, Clone)]
pub enum Literal {
    Int(i32),
    Float(f32),
    String(String),
    Bool(bool),
}

#[derive(Debug, Clone)]
pub struct Member {
    pub target: Box<Span<Expr>>,
    pub ty: Ty,
}

#[derive(Debug, Clone)]
pub enum Stmt {
    /// An expression, used in place of a statement
    Expr(Span<Expr>),

    /// A record declaration
    RecordDecl {
        name: String,
        fields: Vec<(String, Ty)>,
    },

    EnumDecl {
        name: String,
        variants: Vec<(String, Option<Ty>)>,
    },

    /// A function declaration
    FnDecl {
        name: String,
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body: Box<Block>,
    },

    /// A `let` statement, not to be confused with a binding
    Let {
        name: String,
        ty: Ty,
        value: Box<Span<Expr>>,
    },
}

#[derive(Debug, Clone)]
pub enum Expr {
    Literal(Literal),
    Ident(String),
    Member(Member),

    Binop(Box<Span<Expr>>, Op, Box<Span<Expr>>),
    Unop(Op, Box<Span<Expr>>),

    /// A function call
    FnCall {
        target: Box<Span<Expr>>,
        args: Vec<Span<Expr>>,
    },

    Lam {
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body: Box<Span<Expr>>,
    },

    If {
        cond: Box<Span<Expr>>,
        then: Box<Span<Expr>>,
        else_: Option<Box<Span<Expr>>>,
    },
}
