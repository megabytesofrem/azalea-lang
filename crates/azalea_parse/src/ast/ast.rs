use super::ast_types::{Enum, Function, Record, RecordExpr, Ty};
use crate::{lexer::Op, span::Span};

/// A block of statements delimited by `do` and `end`
pub type Block = Vec<Span<Stmt>>;

#[derive(Debug, Clone, PartialEq)]
pub enum Literal {
    Int(i32),
    Float(f32),
    String(String),
    Bool(bool),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Member {
    pub target: Box<Span<Expr>>,
    pub name: String,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Stmt {
    /// An expression, used in place of a statement
    Expr(Span<Expr>),

    /// A `let` statement, *not* a binding
    Let {
        name: String,
        ty: Ty,
        value: Box<Span<Expr>>,
    },

    /// A record declaration
    RecordDecl(Record),

    EnumDecl(Enum),

    /// A function declaration
    FnDecl(Function),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Literal(Literal),
    Ident(String),
    MemberAccess(Member),

    BinOp(Box<Span<Expr>>, Op, Box<Span<Expr>>),
    UnOp(Op, Box<Span<Expr>>),

    Record(RecordExpr),

    Array {
        elements: Vec<Span<Expr>>,
    },

    ArrayIndex {
        target: Box<Span<Expr>>,
        index: Box<Span<Expr>>,
    },

    /// A function call: f(x, y)
    FnCall {
        target: Box<Span<Expr>>,
        args: Vec<Span<Expr>>,
    },

    /// A lambda expression: \(x) -> x + 1
    Lam {
        args: Vec<(String, Ty)>,
        return_ty: Ty,
        body: Box<Span<Expr>>,
    },

    If {
        cond: Box<Span<Expr>>,
        then: Box<Span<Expr>>,
        else_: Option<Box<Span<Expr>>>,
    },
}
