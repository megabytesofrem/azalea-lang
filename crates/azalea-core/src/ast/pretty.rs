use crate::ast::ast_types::{RecordExpr, Ty};
use crate::ast::{Expr, Literal, Member};
use crate::lexer::{Op, SourceLoc};

/// Trait for pretty-printing AST nodes, annotated with their source location.
///
/// Do note that this trait is only used for debugging the AST and pretty-printing type information.
/// It is not intended for user-facing output.
pub trait Pretty {
    fn pretty(&self) -> String;
    fn pretty_with_loc(&self, loc: &SourceLoc) -> String;
}

fn op_to_string(op: &Op) -> String {
    match op {
        Op::Add => "+",
        Op::Sub => "-",
        Op::Mul => "*",
        Op::Div => "/",
        Op::DoubleEq => "==",
        Op::Not => "!",
        Op::NotEq => "!=",
        Op::Less => "<",
        Op::Greater => ">",
        Op::LessEq => "<=",
        Op::GreaterEq => ">=",
        Op::Neg => "-",
    }
    .into()
}

/// Pretty-print a node
pub fn pretty<T: Pretty>(item: &T) -> String {
    item.pretty()
}

/// Pretty-print a node with its source location attached
pub fn pretty_with_loc<T: Pretty>(item: &T, loc: &SourceLoc) -> String {
    item.pretty_with_loc(loc)
}

impl Pretty for Ty {
    fn pretty(&self) -> String {
        match self {
            Ty::Int => "Int".to_string(),
            Ty::Float => "Float".to_string(),
            Ty::String => "String".to_string(),
            Ty::Bool => "Bool".to_string(),
            Ty::Unit => "Unit".to_string(),
            Ty::Any => "Any".to_string(),
            Ty::Unresolved => "?".to_string(),

            // A type variable
            Ty::Var(name) => format!("ᵗ{}", name),

            // A type constructor with optional type arguments
            Ty::TypeCons(name, args) => {
                if args.is_empty() {
                    name.clone()
                } else {
                    let args_str: Vec<String> = args.iter().map(|arg| arg.pretty()).collect();
                    format!("{}[{}]", name, args_str.join(", "))
                }
            }
            Ty::Array(ty) => format!("[{}]", ty.pretty()),
            Ty::Fn(func) => {
                let args_str: Vec<String> = func
                    .args
                    .iter()
                    .map(|(name, ty)| {
                        format!(
                            "{}: {}",
                            if !name.is_empty() { name } else { "_" },
                            ty.pretty()
                        )
                    })
                    .collect();

                let return_ty_str = func.return_ty.pretty();
                format!("fn({}) -> {}", args_str.join(", "), return_ty_str)
            }

            // A universally quantified polymorphic type
            Ty::ForAll(vars, ty) => {
                let vars_str: Vec<String> = vars.iter().map(|v| v.to_string()).collect();
                format!("∀{}. {}", vars_str.join(", "), ty.pretty())
            }

            ty => todo!("Not implemented: {:?}", ty),
        }
    }

    fn pretty_with_loc(&self, loc: &SourceLoc) -> String {
        format!("{} @ {}", self.pretty(), loc)
    }
}

impl Pretty for Literal {
    fn pretty(&self) -> String {
        match self {
            Literal::Int(i) => format!("{}", i),
            Literal::Float(f) => format!("{}", f),
            Literal::String(s) => format!("{}", s),
            Literal::Bool(b) => format!("{}", b),
        }
    }

    fn pretty_with_loc(&self, loc: &SourceLoc) -> String {
        format!("{} @ {}", self.pretty(), loc)
    }
}

impl Pretty for Member {
    fn pretty(&self) -> String {
        format!("{}.{}", self.target.target.pretty(), self.name)
    }

    fn pretty_with_loc(&self, loc: &SourceLoc) -> String {
        format!("{} @ {}", self.pretty(), loc)
    }
}

impl Pretty for RecordExpr {
    fn pretty(&self) -> String {
        let fields: Vec<String> = self
            .fields
            .iter()
            .map(|(name, expr)| format!("{}: {}", name, expr.pretty()))
            .collect();
        format!("{} {{ {} }}", self.name, fields.join(", "))
    }

    fn pretty_with_loc(&self, loc: &SourceLoc) -> String {
        format!("{} @ {}", self.pretty(), loc)
    }
}

impl Pretty for Expr {
    fn pretty(&self) -> String {
        match self {
            Expr::Literal(lit) => lit.pretty(),
            Expr::Ident(ident) => ident.to_string(),
            Expr::MemberAccess(member) => member.pretty(),
            Expr::BinOp(left, op, right) => format!(
                "({} {} {})",
                left.target.pretty(),
                op_to_string(op),
                right.target.pretty()
            ),
            Expr::UnOp(op, expr) => format!("({} {})", op_to_string(op), expr.target.pretty()),
            Expr::Record(record) => record.pretty(),
            Expr::Array { elements } => {
                let elems: Vec<String> = elements.iter().map(|e| e.target.pretty()).collect();
                format!("[{}]", elems.join(", "))
            }
            Expr::ArrayIndex { target, index } => {
                format!("{}[{}]", target.target.pretty(), index.target.pretty())
            }
            Expr::FnCall { target, args } => {
                let args_str: Vec<String> = args.iter().map(|arg| arg.target.pretty()).collect();
                format!("{}({})", target.target.pretty(), args_str.join(", "))
            }
            Expr::Lam {
                args,
                return_ty,
                body,
            } => {
                let args_str: Vec<String> = args.iter().map(|arg| arg.clone().0).collect();
                let return_ty_str = format!(" -> {}", return_ty.pretty());
                format!(
                    "fn({}){} = {}",
                    args_str.join(", "),
                    return_ty_str,
                    body.target.pretty()
                )
            }

            e => {
                // For any other expression, we just use the debug format
                format!("{:?}", e)
            }
        }
    }

    fn pretty_with_loc(&self, loc: &SourceLoc) -> String {
        format!("{} @ {}", self.pretty(), loc)
    }
}
