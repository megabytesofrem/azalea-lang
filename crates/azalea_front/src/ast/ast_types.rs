use super::ast::Expr;

#[derive(Debug, Clone)]
pub enum Ty {
    Int,
    Float,
    String,
    Bool,
    Unit,
    User(String),
    Function(Box<Function>),
    Record(Box<Record>),
}

/// Records are semantically equivalent to JavaScript objects
#[derive(Debug, Clone)]
pub struct Record {
    pub name: String,
    pub fields: Vec<(String, Ty)>,
}

#[derive(Debug, Clone)]
pub struct Enum {
    pub name: String,
    pub variants: Vec<(String, Option<Ty>)>,
}

#[derive(Debug, Clone)]
pub struct Function {
    pub name: String,
    pub args: Vec<(String, Ty)>,
    pub return_ty: Ty,

    pub body: Vec<Expr>,
}
