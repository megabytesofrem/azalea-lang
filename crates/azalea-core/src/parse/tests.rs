use insta::assert_snapshot;

use crate::ast::pretty::{Pretty, pretty_with_loc};
use crate::{parse_expr, parse_stmt};

macro_rules! assert_success {
    ($src:expr) => {
        let expr = crate::parse_expr!($src).unwrap();
        let output = format!(
            "SOURCE:\n{}\n\nEXPRESSION:\n{}",
            $src,
            pretty_with_loc(&expr.target, &expr.loc)
        );
        assert_snapshot!(output);
    };
}

macro_rules! assert_error {
    ($src:expr) => {
        let error = crate::parse::tests::expect_error($src, None);
        let output = format!("SOURCE:\n{}\n\nERROR:\n{}", $src, error);
        assert_snapshot!(output);
    };
    ($src:expr, $hint:expr) => {
        let error = crate::parse::tests::expect_error($src, Some($hint));
        let output = format!("SOURCE:\n{}\n\nERROR:\n{}", $src, error);
        assert_snapshot!(output);
    };
}

pub fn expect_error(src: &str, hint: Option<&str>) -> String {
    let tokens = crate::lexer::lex_tokens(src);
    let mut parser = crate::parse::Parser::new(tokens);

    let result = parser.parse_expr();

    match result {
        Ok(_) => panic!("Expected an error, but got a valid expression"),
        Err(e) => {
            let mut report = e.report();

            if let Some(hint) = hint {
                report.hint = Some(hint.to_string());
            }
            report.source = src.to_string();

            report.show_pretty_source()
        }
    }
}

// Test cases
#[test]
fn invalid_identifier() {
    assert_error!(
        "$florbble",
        "Identifiers cannot start with a special character"
    );
    assert_error!(
        "#1abcdef",
        "Identifiers cannot start with a special character"
    );
}

#[test]
fn bad_string_literal() {
    assert_error!("\"Hello, world!", "Unterminated string literal");
    assert_error!(
        "\"Hello, \\xworld\"",
        "Invalid escape sequence in string literal"
    );
}

#[test]
fn literals() {
    assert_success!("42");
    assert_success!("0x123");
    assert_success!("3.14");
    assert_success!("\"Hello, world!\"");
}

#[test]
fn binary_and_unary_expr() {
    assert_success!("!42");
    assert_success!("1 + 2 * 3");
    // assert_success!("(1 + 2) * 3");
}

#[test]
fn array() {
    assert_success!("[1, 2, 3]");
}

#[test]
fn if_expr() {
    let expr = parse_expr!("if 1 == 1 then 42 end");
    println!("{:#?}", expr);

    assert_success!("if 1 == 1 then 42 end");
}

#[test]
fn lambda_expr() {
    assert_success!("\\(x) -> x + 1");
    assert_success!("\\(x, y) -> x + y");
}

#[test]
fn complex_literals() {
    assert_success!("[1, 2, 3]");
    assert_success!(".{ a: 1, b: 2 }");
    assert_success!(".{ a: 1, b: [2, 3] }");
}

#[test]
fn let_stmt() {
    let stmts = [
        "let x = 42",
        "let x: Int = 42",
        "let name: String = \"Charlotte\"",
        "let lottery_numbers = [21, 38, 42, 14]",
    ];

    for stmt in stmts {
        let result = parse_stmt!(stmt);
        let output = format!("SOURCE:\n{}\n\nAST:\n{:#?}", stmt, result);
        assert_snapshot!(output);
    }
}

#[test]
fn for_loop() {
    let src = "for x in [1, 2, 3] do let x = 5 end";
    let result = parse_stmt!(src);
    let output = format!("SOURCE:\n{}\n\nAST:\n{:#?}", src, result);
    assert_snapshot!(output);
}
