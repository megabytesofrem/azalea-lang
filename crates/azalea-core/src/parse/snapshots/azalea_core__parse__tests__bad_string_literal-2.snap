---
source: crates/azalea-core/src/parse/tests.rs
expression: output
---
SOURCE:
"Hello, \xworld"

ERROR:
Error: Expected an expression
  1 > | "Hello, \xworld"
    ^
Hint: Invalid escape sequence in string literal
